File: greenfloor-engine/src/storage/sqlite/reservations.rs
use std::collections::BTreeMap;

use chrono::{DateTime, Utc};
use rusqlite::params;

use crate::error::{SignerError, SignerResult};

use super::{utcnow_iso, SqliteStore};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OfferReservationLeaseRow {
    pub reservation_id: String,
    pub market_id: String,
    pub wallet_id: String,
    pub asset_id: String,
    pub amount: i64,
    pub status: String,
    pub created_at: String,
    pub expires_at: String,
    pub released_at: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct OfferReservationLeaseRequest<'a> {
    pub reservation_id: &'a str,
    pub market_id: &'a str,
    pub wallet_id: &'a str,
    pub requested_amounts: &'a BTreeMap<String, i64>,
    pub available_amounts: &'a BTreeMap<String, i64>,
    pub lease_seconds: i64,
    pub now: Option<DateTime<Utc>>,
}

impl SqliteStore {
    pub fn try_acquire_offer_reservation_lease(
        &self,
        request: &OfferReservationLeaseRequest<'_>,
    ) -> SignerResult<Option<String>> {
        let OfferReservationLeaseRequest {
            reservation_id,
            market_id,
            wallet_id,
            requested_amounts,
            available_amounts,
            lease_seconds,
            now,
        } = *request;
        if reservation_id.trim().is_empty() {
            return Err(SignerError::Other("reservation_id is required".to_string()));
        }
        if lease_seconds <= 0 {
            return Err(SignerError::Other("lease_seconds must be > 0".to_string()));
        }
        let normalized_requests: BTreeMap<String, i64> = requested_amounts
            .iter()
            .filter_map(|(asset_id, amount)| {
                let amount = *amount;
                (amount > 0).then_some((asset_id.trim().to_ascii_lowercase(), amount))
            })
            .collect();
        if normalized_requests.is_empty() {
            return Ok(Some("reservation_empty_request".to_string()));
        }
        let normalized_available: BTreeMap<String, i64> = available_amounts
            .iter()
            .filter_map(|(asset_id, amount)| {
                let amount = *amount;
                (amount > 0).then_some((asset_id.trim().to_ascii_lowercase(), amount))
            })
            .collect();
        let now_dt = now.unwrap_or_else(Utc::now);
        let now_iso = now_dt.to_rfc3339();
        let expires_at_iso = (now_dt + chrono::Duration::seconds(lease_seconds)).to_rfc3339();

        self.conn.execute("BEGIN IMMEDIATE", []).map_err(|err| {
            SignerError::Other(format!("failed to begin reservation transaction: {err}"))
        })?;
        let result = (|| -> SignerResult<Option<String>> {
            self.conn
                .execute(
                    r"
                UPDATE offer_reservation_lease
                SET status = 'expired',
                    released_at = COALESCE(released_at, ?1)
                WHERE status = 'active'
                  AND expires_at <= ?2
                ",
                    params![now_iso, now_iso],
                )
                .map_err(|err| {
                    SignerError::Other(format!("failed to expire stale reservation leases: {err}"))
                })?;
            let mut stmt = self
                .conn
                .prepare(
                    r"
                SELECT asset_id, COALESCE(SUM(amount), 0) AS reserved_amount
                FROM offer_reservation_lease
                WHERE wallet_id = ?1
                  AND status = 'active'
                  AND expires_at > ?2
                GROUP BY asset_id
                ",
                )
                .map_err(|err| {
                    SignerError::Other(format!("failed to prepare reserved amounts query: {err}"))
                })?;
            let mut rows = stmt.query(params![wallet_id, now_iso]).map_err(|err| {
                SignerError::Other(format!("failed to query reserved amounts: {err}"))
            })?;
            let mut reserved_by_asset: BTreeMap<String, i64> = BTreeMap::default();
            while let Some(row) = rows.next().map_err(|err| {
                SignerError::Other(format!("failed to read reserved amount row: {err}"))
            })? {
                let asset_id: String = row
                    .get(0)
                    .map_err(|err| SignerError::Other(format!("failed to read asset_id: {err}")))?;
                let reserved: i64 = row.get(1).map_err(|err| {
                    SignerError::Other(format!("failed to read reserved_amount: {err}"))
                })?;
                reserved_by_asset.insert(asset_id.trim().to_ascii_lowercase(), reserved);
            }
            for (asset_id, amount) in &normalized_requests {
                let available = normalized_available.get(asset_id).copied().unwrap_or(0);
                let already_reserved = reserved_by_asset.get(asset_id).copied().unwrap_or(0);
                if available - already_reserved < *amount {
                    return Ok(Some(format!(
                        "reservation_insufficient_{asset_id}:available={available}:reserved={already_reserved}:needed={amount}"
                    )));
                }
            }
            for (asset_id, amount) in &normalized_requests {
                self.conn.execute(
                    r"
                    INSERT INTO offer_reservation_lease
                      (reservation_id, market_id, wallet_id, asset_id, amount, status, created_at, expires_at, released_at)
                    VALUES (?1, ?2, ?3, ?4, ?5, 'active', ?6, ?7, NULL)
                    ",
                    params![
                        reservation_id,
                        market_id,
                        wallet_id,
                        asset_id,
                        amount,
                        now_iso,
                        expires_at_iso,
                    ],
                )
                .map_err(|err| {
                    SignerError::Other(format!("failed to insert reservation lease: {err}"))
                })?;
            }
            Ok(None)
        })();
        match result {
            Ok(value) => {
                self.conn.execute("COMMIT", []).map_err(|err| {
                    SignerError::Other(format!("failed to commit reservation transaction: {err}"))
                })?;
                Ok(value)
            }
            Err(err) => {
                let _ = self.conn.execute("ROLLBACK", []);
                Err(err)
            }
        }
    }

    pub fn release_offer_reservation_lease(
        &self,
        reservation_id: &str,
        release_status: &str,
    ) -> SignerResult<u64> {
        let released_at = Utc::now().to_rfc3339();
        let changed = self
            .conn
            .execute(
                r"
                UPDATE offer_reservation_lease
                SET status = ?1, released_at = ?2
                WHERE reservation_id = ?3
                  AND status = 'active'
                ",
                params![release_status, released_at, reservation_id],
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to release reservation lease: {err}"))
            })?;
        super::sqlite_rows_changed(changed)
    }

    pub fn expire_offer_reservation_leases(&self, now: Option<DateTime<Utc>>) -> SignerResult<u64> {
        let now_iso = now.unwrap_or_else(Utc::now).to_rfc3339();
        let changed = self
            .conn
            .execute(
                r"
                UPDATE offer_reservation_lease
                SET status = 'expired',
                    released_at = COALESCE(released_at, ?1)
                WHERE status = 'active'
                  AND expires_at <= ?2
                ",
                params![now_iso, now_iso],
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to expire reservation leases: {err}"))
            })?;
        super::sqlite_rows_changed(changed)
    }

    pub fn add_offer_reservation_lease(
        &self,
        reservation_id: &str,
        market_id: &str,
        wallet_id: &str,
        asset_amounts: &BTreeMap<String, i64>,
        lease_seconds: i64,
    ) -> SignerResult<()> {
        if reservation_id.trim().is_empty() {
            return Err(SignerError::Other("reservation_id is required".to_string()));
        }
        if lease_seconds <= 0 {
            return Err(SignerError::Other("lease_seconds must be > 0".to_string()));
        }
        let created_at = utcnow_iso();
        let expires_at = (Utc::now() + chrono::Duration::seconds(lease_seconds)).to_rfc3339();
        let rows: Vec<(&String, i64)> = asset_amounts
            .iter()
            .filter_map(|(asset_id, amount)| (*amount > 0).then_some((asset_id, *amount)))
            .collect();
        if rows.is_empty() {
            return Err(SignerError::Other(
                "asset_amounts must contain positive amounts".to_string(),
            ));
        }
        for (asset_id, amount) in rows {
            self.conn
                .execute(
                    r"
                    INSERT INTO offer_reservation_lease
                      (reservation_id, market_id, wallet_id, asset_id, amount, status, created_at, expires_at, released_at)
                    VALUES (?1, ?2, ?3, ?4, ?5, 'active', ?6, ?7, NULL)
                    ",
                    params![
                        reservation_id,
                        market_id,
                        wallet_id,
                        asset_id,
                        amount,
                        created_at,
                        expires_at,
                    ],
                )
                .map_err(|err| {
                    SignerError::Other(format!("failed to insert reservation lease: {err}"))
                })?;
        }
        Ok(())
    }

    pub fn list_offer_reservation_leases(
        &self,
        reservation_id: Option<&str>,
    ) -> SignerResult<Vec<OfferReservationLeaseRow>> {
        let mut sql = String::from(
            r"
            SELECT reservation_id, market_id, wallet_id, asset_id, amount, status, created_at, expires_at, released_at
            FROM offer_reservation_lease
            ",
        );
        let mut params: Vec<Box<dyn rusqlite::ToSql>> = Vec::new();
        if let Some(reservation_id) = reservation_id
            .map(str::trim)
            .filter(|value| !value.is_empty())
        {
            sql.push_str("WHERE reservation_id = ?1 ");
            params.push(Box::new(reservation_id.to_string()));
        }
        sql.push_str("ORDER BY id ASC");
        let param_refs: Vec<&dyn rusqlite::ToSql> =
            params.iter().map(std::convert::AsRef::as_ref).collect();
        let mut stmt = self.conn.prepare(&sql).map_err(|err| {
            SignerError::Other(format!("failed to prepare reservation lease query: {err}"))
        })?;
        let mut rows = stmt.query(param_refs.as_slice()).map_err(|err| {
            SignerError::Other(format!("failed to query reservation leases: {err}"))
        })?;
        let mut out = Vec::new();
        while let Some(row) = rows.next().map_err(|err| {
            SignerError::Other(format!("failed to read reservation lease row: {err}"))
        })? {
            out.push(OfferReservationLeaseRow {
                reservation_id: row.get(0).map_err(|err| {
                    SignerError::Other(format!("failed to read reservation_id: {err}"))
                })?,
                market_id: row.get(1).map_err(|err| {
                    SignerError::Other(format!("failed to read market_id: {err}"))
                })?,
                wallet_id: row.get(2).map_err(|err| {
                    SignerError::Other(format!("failed to read wallet_id: {err}"))
                })?,
                asset_id: row
                    .get(3)
                    .map_err(|err| SignerError::Other(format!("failed to read asset_id: {err}")))?,
                amount: row
                    .get(4)
                    .map_err(|err| SignerError::Other(format!("failed to read amount: {err}")))?,
                status: row
                    .get(5)
                    .map_err(|err| SignerError::Other(format!("failed to read status: {err}")))?,
                created_at: row.get(6).map_err(|err| {
                    SignerError::Other(format!("failed to read created_at: {err}"))
                })?,
                expires_at: row.get(7).map_err(|err| {
                    SignerError::Other(format!("failed to read expires_at: {err}"))
                })?,
                released_at: row.get(8).ok(),
            });
        }
        Ok(out)
    }

    pub fn get_offer_reserved_amounts_by_asset(
        &self,
        wallet_id: &str,
    ) -> SignerResult<BTreeMap<String, i64>> {
        let now_iso = utcnow_iso();
        let mut stmt = self
            .conn
            .prepare(
                r"
                SELECT asset_id, COALESCE(SUM(amount), 0) AS reserved_amount
                FROM offer_reservation_lease
                WHERE wallet_id = ?1
                  AND status = 'active'
                  AND expires_at > ?2
                GROUP BY asset_id
                ",
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to prepare reserved amounts query: {err}"))
            })?;
        let mut rows = stmt.query(params![wallet_id, now_iso]).map_err(|err| {
            SignerError::Other(format!("failed to query reserved amounts: {err}"))
        })?;
        let mut out = BTreeMap::default();
        while let Some(row) = rows.next().map_err(|err| {
            SignerError::Other(format!("failed to read reserved amount row: {err}"))
        })? {
            let asset_id: String = row
                .get(0)
                .map_err(|err| SignerError::Other(format!("failed to read asset_id: {err}")))?;
            let amount: i64 = row.get(1).map_err(|err| {
                SignerError::Other(format!("failed to read reserved_amount: {err}"))
            })?;
            out.insert(asset_id, amount);
        }
        Ok(out)
    }

    pub fn prune_offer_reservation_leases(&self, older_than: DateTime<Utc>) -> SignerResult<u64> {
        let cutoff_iso = older_than.to_rfc3339();
        let changed = self
            .conn
            .execute(
                r"
                DELETE FROM offer_reservation_lease
                WHERE status != 'active'
                  AND COALESCE(released_at, expires_at) < ?1
                ",
                params![cutoff_iso],
            )
            .map_err(|err| {
                SignerError::Other(format!("failed to prune reservation leases: {err}"))
            })?;
        super::sqlite_rows_changed(changed)
    }
}

File: tests/helpers/sqlite_store/reservations.py
from __future__ import annotations

from datetime import UTC, datetime

from tests.helpers.sqlite_store.base import SqliteStoreMixin
from tests.helpers.sqlite_store.schema import utcnow_iso


class ReservationStoreMixin(SqliteStoreMixin):

    def add_offer_reservation_lease(
        self,
        *,
        reservation_id: str,
        market_id: str,
        wallet_id: str,
        asset_amounts: dict[str, int],
        lease_seconds: int,
    ) -> None:
        if not reservation_id:
            raise ValueError("reservation_id is required")
        if lease_seconds <= 0:
            raise ValueError("lease_seconds must be > 0")
        if not asset_amounts:
            raise ValueError("asset_amounts must be non-empty")
        created_at = utcnow_iso()
        expires_at = datetime.now(UTC).timestamp() + float(lease_seconds)
        expires_at_iso = datetime.fromtimestamp(expires_at, UTC).isoformat()
        rows = [
            (
                reservation_id,
                market_id,
                wallet_id,
                str(asset_id),
                int(amount),
                "active",
                created_at,
                expires_at_iso,
                None,
            )
            for asset_id, amount in asset_amounts.items()
            if int(amount) > 0
        ]
        if not rows:
            raise ValueError("asset_amounts must contain positive amounts")
        self.conn.executemany(
            """
            INSERT INTO offer_reservation_lease
              (reservation_id, market_id, wallet_id, asset_id, amount, status, created_at, expires_at, released_at)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            rows,
        )
        self.conn.commit()

    def try_acquire_offer_reservation_lease(
        self,
        *,
        reservation_id: str,
        market_id: str,
        wallet_id: str,
        requested_amounts: dict[str, int],
        available_amounts: dict[str, int],
        lease_seconds: int,
        now: datetime | None = None,
    ) -> str | None:
        if not reservation_id:
            raise ValueError("reservation_id is required")
        if lease_seconds <= 0:
            raise ValueError("lease_seconds must be > 0")
        normalized_requests = {
            str(asset_id).strip().lower(): int(amount)
            for asset_id, amount in requested_amounts.items()
            if int(amount) > 0
        }
        if not normalized_requests:
            return "reservation_empty_request"
        normalized_available = {
            str(asset_id).strip().lower(): int(amount)
            for asset_id, amount in available_amounts.items()
            if int(amount) > 0
        }
        now_dt = now or datetime.now(UTC)
        now_iso = now_dt.isoformat()
        expires_at_iso = datetime.fromtimestamp(
            now_dt.timestamp() + float(lease_seconds), UTC
        ).isoformat()
        created_at_iso = now_iso
        try:
            self.conn.execute("BEGIN IMMEDIATE")
            self.conn.execute(
                """
                UPDATE offer_reservation_lease
                SET status = 'expired',
                    released_at = COALESCE(released_at, ?)
                WHERE status = 'active'
                  AND expires_at <= ?
                """,
                (now_iso, now_iso),
            )
            rows = self.conn.execute(
                """
                SELECT asset_id, COALESCE(SUM(amount), 0) AS reserved_amount
                FROM offer_reservation_lease
                WHERE wallet_id = ?
                  AND status = 'active'
                  AND expires_at > ?
                GROUP BY asset_id
                """,
                (wallet_id, now_iso),
            ).fetchall()
            reserved_by_asset = {
                str(row["asset_id"]).strip().lower(): int(row["reserved_amount"] or 0)
                for row in rows
            }
            for asset_id, amount in normalized_requests.items():
                available = int(normalized_available.get(asset_id, 0))
                already_reserved = int(reserved_by_asset.get(asset_id, 0))
                if available - already_reserved < amount:
                    self.conn.rollback()
                    return (
                        f"reservation_insufficient_{asset_id}:"
                        f"available={available}:reserved={already_reserved}:needed={amount}"
                    )
            insert_rows = [
                (
                    reservation_id,
                    market_id,
                    wallet_id,
                    str(asset_id),
                    int(amount),
                    "active",
                    created_at_iso,
                    expires_at_iso,
                    None,
                )
                for asset_id, amount in normalized_requests.items()
            ]
            self.conn.executemany(
                """
                INSERT INTO offer_reservation_lease
                  (reservation_id, market_id, wallet_id, asset_id, amount, status, created_at, expires_at, released_at)
                VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)
                """,
                insert_rows,
            )
            self.conn.commit()
            return None
        except Exception:
            self.conn.rollback()
            raise

    def release_offer_reservation_lease(
        self,
        *,
        reservation_id: str,
        release_status: str,
    ) -> int:
        released_at = utcnow_iso()
        cur = self.conn.execute(
            """
            UPDATE offer_reservation_lease
            SET status = ?, released_at = ?
            WHERE reservation_id = ?
              AND status = 'active'
            """,
            (release_status, released_at, reservation_id),
        )
        self.conn.commit()
        return int(cur.rowcount or 0)

    def expire_offer_reservation_leases(self, *, now: datetime | None = None) -> int:
        now_iso = (now or datetime.now(UTC)).isoformat()
        cur = self.conn.execute(
            """
            UPDATE offer_reservation_lease
            SET status = 'expired',
                released_at = COALESCE(released_at, ?)
            WHERE status = 'active'
              AND expires_at <= ?
            """,
            (now_iso, now_iso),
        )
        self.conn.commit()
        return int(cur.rowcount or 0)

    def prune_offer_reservation_leases(self, *, older_than: datetime) -> int:
        cutoff_iso = older_than.astimezone(UTC).isoformat()
        cur = self.conn.execute(
            """
            DELETE FROM offer_reservation_lease
            WHERE status != 'active'
              AND COALESCE(released_at, expires_at) < ?
            """,
            (cutoff_iso,),
        )
        self.conn.commit()
        return int(cur.rowcount or 0)

    def get_offer_reserved_amounts_by_asset(self, *, wallet_id: str) -> dict[str, int]:
        rows = self.conn.execute(
            """
            SELECT asset_id, COALESCE(SUM(amount), 0) AS reserved_amount
            FROM offer_reservation_lease
            WHERE wallet_id = ?
              AND status = 'active'
              AND expires_at > ?
            GROUP BY asset_id
            """,
            (wallet_id, utcnow_iso()),
        ).fetchall()
        return {str(row["asset_id"]): int(row["reserved_amount"] or 0) for row in rows}

    def list_offer_reservation_leases(
        self,
        *,
        reservation_id: str | None = None,
        include_inactive: bool = True,
    ) -> list[dict[str, str | int | None]]:
        where_clauses: list[str] = []
        params: list[object] = []
        if reservation_id:
            where_clauses.append("reservation_id = ?")
            params.append(reservation_id)
        if not include_inactive:
            where_clauses.append("status = 'active'")
            where_clauses.append("expires_at > ?")
            params.append(utcnow_iso())
        where_sql = ""
        if where_clauses:
            where_sql = "WHERE " + " AND ".join(where_clauses)
        rows = self.conn.execute(
            f"""
            SELECT reservation_id, market_id, wallet_id, asset_id, amount, status, created_at, expires_at, released_at
            FROM offer_reservation_lease
            {where_sql}
            ORDER BY id ASC
            """,
            params,
        ).fetchall()
        return [
            {
                "reservation_id": str(row["reservation_id"]),
                "market_id": str(row["market_id"]),
                "wallet_id": str(row["wallet_id"]),
                "asset_id": str(row["asset_id"]),
                "amount": int(row["amount"] or 0),
                "status": str(row["status"]),
                "created_at": str(row["created_at"]),
                "expires_at": str(row["expires_at"]),
                "released_at": str(row["released_at"]) if row["released_at"] is not None else None,
            }
            for row in rows
        ]
